src/operator/rcnn_utils.h

```diff
@@ -179,7 +181,105 @@ inline void NonMaximumSuppression(const mshadow::Tensor<cpu, 2> &dets,
+//========================
+namespace mxnet {
+namespace op {
+namespace utils {
+
+inline void _mk_anchors(const mshadow::Tensor<cpu, 1>& ws,
+                const mshadow::Tensor<cpu, 1>& hs,
+                float x_ctr,
+                float y_ctr,
+                size_t offset,
+                mshadow::Tensor<cpu, 2> *out_anchors) {
+  size_t num_anchors = ws.size(0);
+  for (size_t i = 0; i < num_anchors; i++) {
+    size_t index = i + offset*num_anchors;
+    (*out_anchors)[index][0] = x_ctr - 0.5*(ws[i] - 1);
+    (*out_anchors)[index][1] = y_ctr - 0.5*(hs[i] - 1);
+    (*out_anchors)[index][2] = x_ctr + 0.5*(ws[i] - 1);
+    (*out_anchors)[index][3] = y_ctr + 0.5*(hs[i] - 1);
+  }
+}
+
+
+std::map<std::string, float> _find_whctrs(const mshadow::Tensor<cpu, 1>&
+                                          anchor) {
+  std::map<std::string, float> whctrs;
+  whctrs["w"] = anchor[2] - anchor[0] + 1;
+  whctrs["h"] = anchor[3] - anchor[1] + 1;
+  whctrs["x_ctr"]  = anchor[0] + 0.5 * (whctrs["w"] -1);
+  whctrs["y_ctr"] = anchor[1] + 0.5 * (whctrs["h"] - 1);
+  return whctrs;
+}
+
+inline void _scale_enum(const mshadow::Tensor<cpu, 2>& ratio_anchors,
+                const mshadow::Tensor<cpu, 1>& scales,
+                mshadow::Tensor<cpu, 2> *scales_tempspace,
+                mshadow::Tensor<cpu, 2> *out_anchors) {
+  std::map<std::string, float> whctrs;
+  mshadow::Tensor<cpu, 1> hs = (*scales_tempspace)[0];
+  mshadow::Tensor<cpu, 1> ws = (*scales_tempspace)[1];
+
+  for (size_t i = 0; i < ratio_anchors.size(0) ; i++) {
+    whctrs = _find_whctrs(ratio_anchors[i]);
+    ws = whctrs["w"] * scales;
+    hs = whctrs["h"] * scales;
+
+    size_t offset = i;
+    _mk_anchors(ws, hs, whctrs["x_ctr"], whctrs["y_ctr"], offset, out_anchors);
+  }
+}
+
+
+inline void _ratio_enum(const mshadow::Tensor<cpu, 1>& base_anchor,
+                const mshadow::Tensor<cpu, 1>& ratios,
+                mshadow::Tensor<cpu, 2> *ratios_tempspace,
+                mshadow::Tensor<cpu, 2> *out_ratio_anchors ) {
+  std::map<std::string, float> whctrs = _find_whctrs(base_anchor);
+  float size = whctrs["w"] * whctrs["h"];
+  mshadow::Tensor<cpu, 1> size_ratios = (*ratios_tempspace)[0];
+  mshadow::Tensor<cpu, 1> ws = (*ratios_tempspace)[1];
+  mshadow::Tensor<cpu, 1> hs = (*ratios_tempspace)[2];
+
+  size_ratios = size/ratios;
+  for (size_t i = 0; i < ratios.size(0); i++) {
+    ws[i] = std::round(std::sqrt(size_ratios[i]));
+    hs[i] = std::round(ws[i] * ratios[i]);
+  }
+  size_t offset = 0;
+  _mk_anchors(ws, hs, whctrs["x_ctr"], whctrs["y_ctr"], offset,
+              out_ratio_anchors);
+}
+
+inline void  generate_anchors(int base_size, const <cpu, 1>& ratios,
+                       const mshadow::Tensor<cpu, 1>& scales,
+                       mshadow::TensorContainer<cpu, 2>* out_anchors) {
+  // Construct base anchor
+  float base_anchor_data[] = {0, 0, base_size - 1, base_size -1};
+  mshadow::TensorContainer<cpu, 1> base_anchor(Shape1(4));
```

Comment:
Remove all TensorContainers in side of utility function. If you need these space, try to pass from outside